monads/test/Spec.hs

```diff
@@ -0,0 +1,106 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+import Hedgehog
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+import Test.Tasty
+import Test.Tasty.Hedgehog
+
+import qualified Data.Map as M
+
+import HW.StackMachine
+import HW.Eval
+import HW.Compiler
+import Reader.Eval
+import Reader.MyReader
+import Expr
+
+-- Helper function to evaluate an expression with a custom context
+evalWithContext :: (Ord a) => Expr a -> M.Map a Int -> Int
+evalWithContext expr context = Reader.MyReader.runMyReader (Reader.Eval.eval expr) context
+
+-- Generate an integer within a range
+genInt :: Gen Int
+genInt = Gen.int (Range.constant 0 100)
+
+-- Generate expressions using only predefined variables from the initial context
+genExpr :: Int -> [String] -> Gen (Expr String)
+genExpr 0 boundVars =
+  if null boundVars
+    then Num <$> genInt
+    else Gen.choice [Num <$> genInt, Expr.Var <$> Gen.element boundVars]
+
+genExpr n boundVars = do
```

Comment:
It's better to use `Gen.recursive`: https://hackage.haskell.org/package/hedgehog-1.5/docs/Hedgehog-Gen.html#v:recursive

Reply:
Fixed.